### DISCUSSION TITLE
Set picocli's version at build time using the new //MANIFEST functionality

### DISCUSSION THREAD

**Discussion by helpermethod** [jbangdev/jbang#1455] [Show and tell]
With the new //MANIFEST support, it's now very easy to set the version of a JBang picocli application at build time ❤️ !

Here are the necessary steps

Set the `Implementation-Version` to `${version}` using the new //MANIFEST support

    //MANIFEST Implementation-Version=${version}

Implement an `IVersionProvider` which reads the version from the manifest

```java
static class ManifestVersionProvider implements IVersionProvider {
    public String[] getVersion() {
       return new String[] { getClass().getPackage().getImplementationVersion() };
    }
 }
```

Reference the `versionProvider` in picocli's `@Command` annotation

```java
@Command(
    name = "p2e",
    mixinStandardHelpOptions = true,
    versionProvider = p2e.ManifestVersionProvider.class,
    description = "Convert Spring configuration property names to environment variable names."
)
```

Add a package (otherwise the MANIFEST.MF will not be found)

```java
package com.github.helpermethod.p2e;
```

You can now set the `version` at build time!

    jbang -Dversion=6.6.6 p2e.java --version

Works in the CI too

https://github.com/helpermethod/p2e/blob/main/.github/workflows/release.yml

**Comment by quintesse** [comment]
Thanks for  providing an example of the new feature!  👍 🙂

**Reply by maxandersen** [reply]
In case you want it to work without always setting version you can do:

`...=${version:dev}`

**Reply by helpermethod** [reply]
Good point!
